experiment/tor/tor.go

```diff
@@ -0,0 +1,241 @@
+// Package tor contains the tor experiment.
+//
+// Spec: https://github.com/ooni/spec/blob/master/nettests/ts-023-tor.md
+package tor
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ooni/probe-engine/experiment"
+	"github.com/ooni/probe-engine/experiment/handler"
+	"github.com/ooni/probe-engine/experiment/httpheader"
+	"github.com/ooni/probe-engine/internal/netxlogger"
+	"github.com/ooni/probe-engine/internal/oonidatamodel"
+	"github.com/ooni/probe-engine/internal/oonitemplates"
+	testlists "github.com/ooni/probe-engine/internal/orchestra/testlists/tor"
+	"github.com/ooni/probe-engine/model"
+	"github.com/ooni/probe-engine/session"
+)
+
+const (
+	testName    = "tor"
+	testVersion = "0.0.1"
+)
+
+// Config contains the experiment config.
+type Config struct{}
+
+// TargetResults contains the results of a target.
+type TargetResults struct {
+	Agent          string                          `json:"agent"`
+	Failure        *string                         `json:"failure"`
+	NetworkEvents  oonidatamodel.NetworkEventsList `json:"network_events"`
+	Queries        oonidatamodel.DNSQueriesList    `json:"queries"`
+	Requests       oonidatamodel.RequestList       `json:"requests"`
+	TargetAddress  string                          `json:"target_address"`
+	TargetProtocol string                          `json:"target_protocol"`
+	TCPConnect     oonidatamodel.TCPConnectList    `json:"tcp_connect"`
+	TLSHandshakes  oonidatamodel.TLSHandshakesList `json:"tls_handshakes"`
+}
+
+// TestKeys contains tor test keys.
+type TestKeys struct {
+	Targets []TargetResults `json:"targets"`
+}
+
+type measurer struct {
+	config Config
+}
+
+func newMeasurer(config Config) *measurer {
+	return &measurer{config: config}
+}
+
+func (m *measurer) measure(
+	origCtx context.Context,
+	sess *session.Session,
+	measurement *model.Measurement,
+	callbacks handler.Callbacks,
+) error {
+	ctx, cancel := context.WithTimeout(origCtx, 60*time.Second)
+	defer cancel()
+	// fetch experiment targets
+	targets, err := testlists.Query(ctx, testlists.Config{})
+	if err == nil {
+		// measure targets
+		err = m.measureTargets(origCtx, sess, measurement, callbacks, targets)
+	}
+	return err
+}
+
+func (m *measurer) measureTargets(
+	ctx context.Context,
+	sess *session.Session,
+	measurement *model.Measurement,
+	callbacks handler.Callbacks,
+	targets []model.TorTarget,
+) error {
+	// run measurements in parallel
+	var waitgroup sync.WaitGroup
+	rc := newResultsCollector(sess, measurement, callbacks)
+	waitgroup.Add(len(targets))
+	workch := make(chan model.TorTarget)
+	const parallelism = 2
```

Comment:
Perhaps we should define the parallelism constant higher up inside of this file with a comment above it saying that we will be measuring two targets at the same time?

Reply:
Sure

Reply:
Sure!

Reply:
done!